### ISSUE TITLE
RFC Implement a build targeted at detecting performance regression

### ISSUE THREAD

**Issue by maxime-rainville** [silverstripe/silverstripe-framework#9534]
## Description

Our current process has difficulty detecting performance regression. We might implement a change in a specific context, but neglect to test it with a large number of files/block/pages.

We could implement a build that run on a schedule that runs a few common task that have a tendency to be slow. The build would measure the time it takes to run each task and if it falls outside a reasonable range the build would fail.

We could also get the execution times uploaded somewhere so we could graph it over time.

**Comment by maxime-rainville** [comment]
Here's an example of possible tasks that could be included in this build:
- Accessing `/admin/pages/edit/treeview` on a site with lots of pages
- Renaming a folder with lots of sub-files/sub-folders
- Creating lots of pages with `sake dev/tasks/FTPageMakerTask`
- Publishing a large campaign
- Rendering an Elemental Area with lots of blocks

**Comment by michalkleiner** [comment]
That sounds like a good idea :)

**Comment by blueo** [comment]
good idea :+1: blackfire seems to do some open source collaboration https://blog.blackfire.io/sponsoring-open-source-projects.html and i've used Tideway's open source xhprof extension before with success: https://github.com/tideways/php-xhprof-extension.

**Comment by mfendeksilverstripe** [comment]
Great to see this :)

Some additional areas to track:

- `TreeDropDownField`
- CMS page search with filtering

**Comment by chillu** [comment]
Great idea. I think we should focus on most common tasks, rather than "tasks which we expect to be slow". But all of this should be done on a realistic site (thousands of pages with blocks, thousands of files). 

Blackfire would be fantastic, since we are not as dependant on execution performance in Travis VMs - we can count method invocations in the call stack. Max, do you want to contact them and see what the conditions for a sponsorship would be? Please CC me on this.

That being said, Symfony does this a lot simpler - they just set max running times on granular method-level performance tests: https://github.com/symfony/symfony/blob/331765b95df3c10ceb4b4162b86c0a1ac3cb6154/src/Symfony/Component/Form/Test/FormPerformanceTestCase.php#L42. It doesn't [look like](https://travis-ci.org/github/symfony/symfony/jobs/694841385) they execute those tests in CI though. And it's pretty coarse, they just complain when a specific invocation takes longer than a second in many cases. Blackfire would give us the ability to compare runs to a baseline stored within their service. We could also achieve this by storing e.g. JSON in S3, and aggregating runtimes of tests. It's a bit like code coverage though: The aggregate might not be much worse, but you've actually made a crucial method 10x slower.

Has anyone seen PHP projects which do automated performance regression testing in CI?

**Comment by blueo** [comment]
If the blackfire thing doens't work out then here's an example of reading the xhprof file from php https://github.com/sugarcrm/xhprof-viewer you could use that kind of thing for counting method calls etc
